File: cardmasters_app/cardmasters_app/report/production_lifecycle_report/production_lifecycle_report.py
# Copyright (c) 2025, Shan Torrejos and contributors
# For license information, please see license.txt

# import frappe

import frappe

def execute(filters=None):
    columns = [
        {"label":"Sales Order", "fieldname":"sales_order", "fieldtype":"Link", "options":"Sales Order", "width":200},
        {"label":"Work Order",  "fieldname":"work_order",  "fieldtype":"Link", "options":"Work Order",  "width":200},
        {"label":"Job Card",    "fieldname":"job_card",    "fieldtype":"Link", "options":"Job Card",    "width":200},
    ]

    data = []
    for so in frappe.get_all("Sales Order", fields=["name","status"]):
        first_so_row = True
        work_orders = frappe.get_all("Work Order", fields=["name","status"], filters={"sales_order": so.name})
        for wo in work_orders:
            job_cards = frappe.get_all("Job Card", fields=["name","status"], filters={"work_order": wo.name})
            for idx, jc in enumerate(job_cards):
                row = {}
                # only show SO name & status on the very first JC row
                if first_so_row and idx == 0:
                    row["sales_order"] = so.name
                    row["sales_order_status"] = so.status
                    first_so_row = False
                # only show WO name & status on its first JC
                if idx == 0:
                    row["work_order"] = wo.name
                    row["work_order_status"] = wo.status
                # always show JC name & status
                row["job_card"] = jc.name
                row["job_card_status"] = jc.status

                data.append(row)

    return columns, data

File: cardmasters_app/cardmasters_app/report/production_lifecycle_report/production_lifecycle_report.js
// Copyright (c) 2025, Shan Torrejos and contributors
// For license information, please see license.txt

frappe.query_reports["Sales ➔ Work ➔ Job Card"] = {
	formatter: function(value, row, column, data, default_formatter) {
	  // base rendering (including the link)
	  let v = default_formatter(value, row, column, data);
  
	  // Sales Order: add status if present
	  if (column.fieldname === "sales_order" && data.sales_order_status) {
		v = `${v} (${data.sales_order_status})`;
	  }
  
	  // Work Order: indent when SO is blank, then add status
	  if (column.fieldname === "work_order") {
		let pad = !data.sales_order ? "padding-left:1.5em;" : "";
		let txt = v;
		if (data.work_order_status) {
		  txt = `${txt} (${data.work_order_status})`;
		}
		return `<div style="${pad}">${txt}</div>`;
	  }
  
	  // Job Card: always indented deeper + add status
	  if (column.fieldname === "job_card") {
		let txt = v;
		if (data.job_card_status) {
		  txt = `${txt} (${data.job_card_status})`;
		}
		return `<div style="padding-left:3em;">${txt}</div>`;
	  }
  
	  return v;
	}
  };
  